runtime/src/accounts_db.rs

```diff
@@ -259,7 +267,7 @@ impl AccountStorageEntry {
     }
 
     pub fn slot(&self) -> Slot {
-        self.slot
+        self.slot.load(Ordering::Acquire)
```

Comment:
👀

Reply:
Our first acquire-release semantics! To celebrate maybe we should include a comment about why it's necessary here?

Reply:
I think we use it in gossip as well. Most of our atomics are not critical for functionality like perf counters which are fine if they are incremented or loaded out of order. I thought it would be bad if the load/store is re-ordered for the slot value, a thread could see a stale slot value.

Reply:
Just to clarify my understanding and future reference about the interaction between the atomics `slot` and `id` fields in `AccountStorageEntry`:

A recycle scenario can be summarized with something like the following:

Initial state: 
```
self.slot = 0
self.id = 0
```

Thread A grabs and holds a copy of the storage entry `Arc` and reads
`Thread A`
```
let ref_count = storage.lookup(store_id_x);  // lookup in the storage for some store_id_x, may be the old store id, or the new recycled id
ref_count.store(2, Ordering::SeqCst);      // storage clone bumps the Arc refcount, Arc uses `SeqCst`   
slot.load(Ordering::Acquire);                             
id.load(Ordering::Relaxed);
ref_count.store(1, Ordering::SeqCst);
```

Thread `B` performing the `recycle`:
`Thread B`
```
// do the recycle
if refcount.load(Ordering::SeqCst) == 1 {    // Arc `strong_count()` uses `SeqCst`   
    slot.store(5, Ordering::Release);                             
    id.store(5, Ordering::Relaxed);
}
// Insert the store
storage.lock().insert(store_id_y);   // insert recycled store into storage with new store id
```

So now looking at possible combinations of `(slot, id) ` that thread `A` can see: 

1) If Thread `A` is reading directly grabbing and holding copy of the `Arc` for the storage entry from the `recycled_stores`, then it can see  any combination of (0, 0), (0, 5), (5, 5) (5, 0), because the loads can interleave with the recycle logic in `B` that starts at `if Arc::strong_count(store) == 1 {`

2) If Thread `A` is finding the storage entry through a read on `self.storages`:

Can thread `A` still see a mixed combination like (0, 5), or (5, 0)? I think the storage lock and the `strong_count() == 1` check prevent this from happening?

For example for `(5, 0)`, this means `A` must have seen `B`'s store to `slot`, which means store_id_x == store_id_y,
so the loads in thread `A` must have occurred *after* the new storage with `store_id_y` was inserted
into the store at `storage.lock().insert(store_id_y);`. And because locks have full fences, both loads in 
thread `A` cannot be ordered before the `storage.lock().insert(store_id_y);`, and must see both stores from thread `B`.

`(0, 5)` follows a similar reasoning.